### PR TITLE
fix!: issues discovered during cluster tests

### LINKED ISSUE #50: Fix issues discovered during Cluster Tests
Several issues were discovered during testing on the cluster:

- Add CAPTUM: Occlusion method
- resolve "auto\_pred" target for explainers
- Optimize memory usage in batch attribution processing
- fix SHAP auto\_pred target device mismatch
- SHAP-Visualizer: remove plt.tight\_layout to fix display bug
- align SHAP visualizer with CAPTUM to display title-info for each image

### PR DESCRIPTION
**Pull request by pizzajojo**
**Summary**

**Configuration & Documentation Enhancements:**

- Added an archived, comprehensive configuration guide that explains Hydra-based overrides, config groups, background data for SHAP, progress bar options, sample name visualisation, custom data/model support, outputs, batch runs, dry runs, and the Python API.
- Expanded documentation for `ShapImageVisualiser`, detailing its configuration, parameters, and usage examples for consistent visual outputs across explainers.

**Pipeline & Runtime Improvements:**

- Enhanced the pipeline to support `"auto_pred"` as a special value for explainer `target`—automatically passing model predictions as the target for explainers when configured. This includes a helper function and corresponding tests. [\[1\]](diffhunk://#diff-f962283d936c65697c18c204c4ccbb07ffbad27fd22874e34f6d135a4b6f2c29L92-R98) [\[2\]](diffhunk://#diff-f962283d936c65697c18c204c4ccbb07ffbad27fd22874e34f6d135a4b6f2c29R156-R172) [\[3\]](diffhunk://#diff-82beb4eec2578f17433fa93b1284136efa8fea44dcac431ee0ad29b5f75e4513R522-R550) [\[4\]](diffhunk://#diff-dd2c797f3cec0961b47d6fff2d526b4529919ebe71346156bf170f9988044117L100-R100)
- Improved robustness of the `_run_without_tracking` function by safely handling models with or without a `backend` attribute, and ensuring correct input preparation and prediction extraction. [\[1\]](diffhunk://#diff-f962283d936c65697c18c204c4ccbb07ffbad27fd22874e34f6d135a4b6f2c29L6-R10) [\[2\]](diffhunk://#diff-f962283d936c65697c18c204c4ccbb07ffbad27fd22874e34f6d135a4b6f2c29L92-R98)

**Resource Management & Batch Processing:**

- Added explicit memory cleanup (using `gc.collect()` and `torch.cuda.empty_cache()`) after each batch in `BaseExplainer` to prevent out-of-memory errors during long or large batched runs. [\[1\]](diffhunk://#diff-ca5c8b4ee850b26e6912f3d89580258dde2bfbe8fcedb2e87143eca18be173e6R5) [\[2\]](diffhunk://#diff-ca5c8b4ee850b26e6912f3d89580258dde2bfbe8fcedb2e87143eca18be173e6L107-R120)

**Explainer-Specific Fixes:**

- Ensured that for the Captum `Occlusion` algorithm, `sliding_window_shapes` and `strides` are always passed as tuples, not lists, to comply with Captum's API requirements. [\[1\]](diffhunk://#diff-1d0909b55de0e5fa15c38e5aaa7db02b1d89db2babc08d6fa6abc798427907dcR111-R113) [\[2\]](diffhunk://#diff-1d0909b55de0e5fa15c38e5aaa7db02b1d89db2babc08d6fa6abc798427907dcR138-R151)

**Cluster/Launcher Support:**

- Added optional dependency and configuration for Hydra's `submitit` SLURM launcher, with a sample config for single-GPU jobs. [\[1\]](diffhunk://#diff-50c86b7ed8ac2cf95bd48334961bf0530cdc77b5a56f852c5c61b89d735fd711R100-R102) [\[2\]](diffhunk://#diff-dff04a87d35d186cfca3b8a062ced46fa46120446d6c0e4cf8d633e68403b6a7R1-R30)

## Checklist

- [ ] **CI** — Required checks are green
- [ ] **Breaking changes** — If this PR breaks compatibility (API, configs, file formats, etc.), the title uses `!` after the type or scope (e.g. `feat!:` or `feat(transparency)!:`). I understand that this change will have a **direct, disruptive impact** on package users. I ensured that **this change is absolutely necessary and cannot be delayed**.
- [ ] **Contributor guide** — I’ve read [**Pull requests and commit messages**](https://caiivs.github.io/raitap/contributor/pull-requests.html) before requesting review.

## Optional

- [x] **Issue** — A GitHub issues is linked to this PR ("Development" section in the right sidebar).
- [ ] **Docs** — User or contributor docs updated where needed.
- [ ] **Tests** — New or updated tests cover the change.
  This pull request introduces several improvements and new features to RAITAP, focusing on enhanced configuration flexibility, improved SHAP visualisation options, more robust pipeline execution, and better resource management during batched explanations. It also adds cluster/launcher support via Hydra and improves documentation for configuration and usage.

### CODE REVIEW

File: src/raitap/transparency/tests/e2e_case_matrix.py [hunk 1]
```diff
@@ -529,13 +538,34 @@ def render_mpl_figure(case: MatrixCase, request: pytest.FixtureRequest) -> Figur
...
-        model=model,
-        inputs=inputs,
-        target=target,
-        background=background,
-    )
+
+    if case.mode == "explain" and case.explain_call_kwargs:
+        # Route through _run_explain_case so explain_call_kwargs (e.g. Occlusion's
+        # sliding_window_shapes / strides) are forwarded automatically.  We only read
+        # explanation.attributions and render the figure ourselves; explanation.visualise()
```
Comment (stanlrt) [inline src/raitap/transparency/tests/e2e_case_matrix.py]: Doesnt that completely defeat the point of e2e testing visulasiers? we should play with pytest-mpl's tolerance instead

File: src/raitap/transparency/tests/e2e_case_matrix.py [hunk 2]
```diff
@@ -609,6 +639,39 @@ def render_mpl_figure(case: MatrixCase, request: pytest.FixtureRequest) -> Figur
...
+        mode="explain",
+        algorithm="Occlusion",
+        needs_fixture="needs_captum",
+        model_fixture="simple_cnn",
+        input_fixture="sample_images",
+        target_mode=0,
+        expected_shape="same_as_inputs",
+        run_dir_mode="plain_transparency",
+        explain_call_kwargs={
+            # Use a sub-image occlusion window so the deterministic MPL baseline
```
Comment (stanlrt) [inline src/raitap/transparency/tests/e2e_case_matrix.py:652]: what does that mean?

File: src/raitap/transparency/visualisers/shap_visualisers.py [hunk 3]
```diff
@@ -258,21 +299,55 @@ class ShapImageVisualiser(BaseVisualiser):
 
     compatible_algorithms: frozenset[str] = frozenset({"GradientExplainer", "DeepExplainer"})
 
-    def __init__(self, max_samples: int = 4):
+    def __init__(
+        self,
```
[the remark this answers is not in the document]
Comment (stanlrt) [reply src/raitap/transparency/visualisers/shap_visualisers.py:320]: @pizzajojo yeah we either need to update the docs to explain why we dont rely on SHAP anymore and how users can customise them

---

This is a response to a Copilot Review comment

[the remark this answers is not in the document]
Comment (stanlrt) [reply src/raitap/transparency/visualisers/shap_visualisers.py:320]: see my latest doc PR (merge it into your branch)

---

This is a response to a Copilot Review comment

File: src/raitap/transparency/results.py [hunk 4]
```diff
@@ -123,6 +123,12 @@ def visualise(self, **kwargs: Any) -> list[VisualisationResult]:
                 visualise_kwargs["sample_names"] = sample_names
             if supports_show_sample_names:
                 visualise_kwargs["show_sample_names"] = show_sample_names
+            if (
+                type(vis).__name__ == "ShapImageVisualiser"
+                and self.algorithm
+                and "title" not in visualise_kwargs
+            ):
+                visualise_kwargs["title"] = f"{self.algorithm} (SHAP)"
```
[the remark this answers is not in the document]
Comment (stanlrt) [reply src/raitap/transparency/results.py]: @pizzajojo yeah thats very fragile

---

This is a response to a Copilot Review comment

File: pyproject.toml [hunk 5]
```diff
@@ -95,6 +95,9 @@ shap = ["shap>=0.46.0"]
 captum = ["captum>=0.7.0"]
 transparency = ["raitap[shap,captum]"]
 
+# Cluster / launcher support
+cluster = ["hydra-submitit-launcher>=1.2.0"]
```
Comment (stanlrt) [inline pyproject.toml]: should we really name it "cluster" if it's useful for any remote server kind stuff? **users might not call it a cluster**, and need docs about this (see **your TODO about Slurm and batch in docs**)

**Review by stanlrt — CHANGES_REQUESTED** [review]

File: src/raitap/transparency/visualisers/shap_visualisers.py [hunk 6]
```diff
@@ -32,6 +32,51 @@ def _close_and_return(fig: Figure) -> Figure:
...
+        return f"{base_title}: {sample_name}"
+    return base_title or sample_name
+
+
+def _resolve_title(
+    *, explicit_title: str | None, fallback_title: str | None, sample_name: str | None = None
+) -> str | None:
+    """Resolve titles while preserving an explicitly provided empty string."""
+    base_title = explicit_title if explicit_title is not None else fallback_title
+    return _compose_title(base_title, sample_name)
```
[the remark this answers is not in the document]
Comment (pizzajojo) [reply src/raitap/transparency/visualisers/shap_visualisers.py:67]: fixed

File: src/raitap/transparency/results.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (pizzajojo) [reply src/raitap/transparency/results.py]: fixed

File: src/raitap/transparency/tests/e2e_case_matrix.py (hunk 2 above)
Comment (pizzajojo) [reply src/raitap/transparency/tests/e2e_case_matrix.py:652]: fixed

File: pyproject.toml (hunk 5 above)
Comment (pizzajojo) [reply pyproject.toml]: yeah I mean that's what submitit is for..

File: src/raitap/transparency/visualisers/shap_visualisers.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (pizzajojo) [reply src/raitap/transparency/visualisers/shap_visualisers.py:320]: added section in the documentation

File: docs/archive/consumers/configuration.md [hunk 7]
```diff
@@ -0,0 +1,279 @@
+# Archived: Configuration guide
+
```
Comment (stanlrt) [inline docs/archive/consumers/configuration.md]: delete this file

**Review by stanlrt — CHANGES_REQUESTED** [review]

File: src/raitap/transparency/explainers/base_explainer.py [hunk 8]
```diff
@@ -104,14 +105,19 @@ def _compute_with_optional_batches(
...
+                **batch_kwargs,
             )
+            chunks.append(chunk.detach().cpu())
+            del chunk, batch_inputs, batch_kwargs
+            # Clean up memory after each batch to avoid OOM errors in long runs
+            # (e.g. with SHAP partition explainer).
+            gc.collect()
+            if torch.cuda.is_available():
+                torch.cuda.empty_cache()
         return torch.cat(chunks, dim=0)
```
[the remark this answers is not in the document]
Comment (stanlrt) [reply src/raitap/transparency/explainers/base_explainer.py:129]: @pizzajojo sounds complicated but important

File: pyproject.toml (hunk 5 above)
Comment (pizzajojo) [reply pyproject.toml]: changed to launcher

File: src/raitap/transparency/explainers/base_explainer.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (pizzajojo) [reply src/raitap/transparency/explainers/base_explainer.py:129]: fixed by always normalize the collected attributions as cpu tensors

File: docs/archive/consumers/configuration.md (hunk 7 above)
Comment (pizzajojo) [reply docs/archive/consumers/configuration.md]: deleted

File: docs/using-raitap/job-launcher.md [hunk 9]
```diff
@@ -0,0 +1,209 @@
...
+
+The simplest approach is to use Hydra's built-in `submitit_slurm` launcher and override parameters directly from the CLI:
+
+```{install-tabs}
+:uv:
+uv run --extra launcher --extra torch-cuda --extra captum --extra shap python \
+  path/to/your_entrypoint.py \
+  --multirun \
+  hydra/launcher=submitit_slurm \
+  transparency=captum_ig,shap_gradient \
```
Comment (stanlrt) [inline docs/using-raitap/job-launcher.md:75]: this looks like an example with arbitrary frameworks, this should be made clear (text before the snippet saying "In the example below" or such a thing)

File: docs/using-raitap/job-launcher.md [hunk 10]
```diff
@@ -0,0 +1,209 @@
...
+If you run sweeps frequently with the same resource settings, create a Hydra launcher preset.
+For example, create `configs/hydra/launcher/my_launcher.yaml`:
+
+```yaml
+# @package hydra.launcher
+defaults:
+  - submitit_slurm
+
+# Job resource defaults
+submitit_folder: ${hydra.sweep.dir}/.submitit/%j
```
Comment (stanlrt) [inline docs/using-raitap/job-launcher.md:114]: i dont understand where this belongs, i dont see it in [schema.py](schema.py). Users will have a single YAML and wont have the `defaults` key, this is internal to RAITAP

See how other config docs show the YAML

File: docs/using-raitap/job-launcher.md [hunk 11]
```diff
@@ -0,0 +1,209 @@
...
+module load python/3.13.2
+
+# Load uv if provided as a module
+VENV=raitap-env module load uv/0.6.12
+```
+
+**3. Sync dependencies:**
+
+```bash
+uv sync --extra launcher --extra torch-cuda --extra captum --extra shap --extra metrics
```
Comment (stanlrt) [inline docs/using-raitap/job-launcher.md]: same here, no sync for consumers

File: docs/using-raitap/job-launcher.md (hunk 9 above)
Comment (stanlrt) [reply docs/using-raitap/job-launcher.md:75]: its an example

File: docs/using-raitap/job-launcher.md (hunk 10 above)
Comment (stanlrt) [reply docs/using-raitap/job-launcher.md:114]:
```
hydra:
  run:
    dir: "./custom-outputs-dir"
  laucnher:
    # Jonas' options
```

**Review by stanlrt — APPROVED** [review]
